# Tensor reduction operations

A reduction is any operation that reduces a tensor to a smaller number of elements. Common examples include:

- sum()
- mean()
- max()

In this lecture, we will explore how the dim argument influences reduction operations.
In short, the dim argument determines along which axis the tensor collapses when a reduction is applied.

Before diving into that, let’s first understand how reduction operations affect a tensor when no dim argument is provided.

In [ ]:
import torch

# A 2×3 tensor representing grades for two persons
# across three tasks
grades = torch.tensor([[20., 25., 30.],
                       [30., 20., 25.]])

# Reduction operations WITHOUT specifying dim:
# These operate on ALL elements in the tensor.

mean_value = grades.mean() # Mean of all numbers
sum_value = grades.sum() # Sum of all numbers
max_value = grades.max() # Highest value in the tensor

print(f"Overall mean: {mean_value}")
print(f"Sum of all values in tensor: {sum_value}")
print(f"Highest value in tensor: {max_value}")


Overall mean: 25.0
Sum of all values in tensor: 150.0
Highest value in tensor: 30.0


 ---
## The **dim** argument for **2D tensors**

When PyTorch performs reduction operations, you can control which direction the tensor collapses. In other words, you decide whether PyTorch should combine values row‑by‑row or column‑by‑column

- no dim argument
    - The entire tensor collapses into a single value.

- dim = 0
    - Collapses the rows, operating vertically. The result is a vector whose length equals the number of columns.

- dim = 1
    - Collapses the columns, operating horizontally. The result is a vector whose length equals the number of rows.

In our code example, the grades tensor is a 2×3 tensor representing the grades of two individuals across three tasks in a course:
[20 25 30
 30 20 25]

Now let’s see how the dim argument changes the outcome of the reduction operations we used earlier.

In [12]:
# Collapses vertically = mean for each task
avg_per_task = age.mean(dim=0)

# Collapses horizontally = mean for each student
avg_per_person = age.mean(dim=1)

print(f"Average grade for each task: {avg_per_task}")
print(f"Average grade for each student: {avg_per_person}")


Average grade for each task: tensor([25.0000, 22.5000, 27.5000])
Average grade for each student: tensor([25., 25.])


**Template of the dim argument**

| Person      | Task1 | Task2 | Task3 | mean(dim=1) |
|-------------|-------|-------|-------|--------------|
| Person 1    | 20    | 25    | 30    | 25           |
| Person 2    | 30    | 20    | 25    | 25           |
| mean(dim=0) | 25.0  | 22.5  | 27.5  | —            |

**Notice the power of the dim argument**
Understanding this argument is essential. You will use it throughout your entire PyTorch workflow, from reductions to batching, loss functions, and neural network outputs.